[red-knot] @override lint rule

carljm opened the pull request:
## Summary

Lots of TODOs and things to clean up here, but it demonstrates the working lint rule.

## Test Plan

```
➜ cat main.py
from typing import override
from base import B

class C(B):
    @override
    def method(self): pass

➜ cat base.py
class B: pass

➜ cat typing.py
def override(func):
    return func
```

(We provide our own `typing.py` since we don't have typeshed vendored or type stub support yet.)

```
➜ ./target/debug/red_knot main.py
...
1   0.012086s TRACE red_knot Main Loop: Tick
[crates/red_knot/src/main.rs:157:21] diagnostics = [
    "Method C.method is decorated with `typing.override` but does not override any base class method",
]
```

If we add `def method(self): pass` to class `B` in `base.py` and run red_knot again, there is no lint error.

AlexWaygood commented on crates/red_knot/src/lint.rs:
```diff
@@ -140,6 +142,66 @@ fn lint_unresolved_imports(context: &SemanticLintContext) -> QueryResult<()> {
 }
 
+fn lint_bad_overrides(context: &SemanticLintContext) -> QueryResult<()> {
+    // TODO possibly we should have a special marker on the real typing module (from typeshed) so
```
Yes, mypy special-cases a small number of stdlib modules (including `typing`) such that if one of them is shadowed, that's a hard error and mypy will refuse to do any type-checking at all. I think it would make sense for us to do similarly. See e.g. https://github.com/python/mypy/pull/13155

We may also want a more general way to be able to distinguish between "`datetime.datetime`, where `datetime` is the stdlib `datetime` module", and "`datetime.datetime`, where `datetime` is a first-party user module" though. For a type checker that distinction doesn't matter so much — as long as it can follow the module resolution and figure out the type, it's all good. But we probably _don't_ want to e.g. emit our `flake8-datetimez` rules on code using functions from random user modules that shadow the stdlib `datetime` module. We probably _only_ want to emit those lines on code using functions imported from the _actual stdlib_ `datetime` module.

AlexWaygood commented on crates/red_knot/src/types.rs:
```diff
@@ -255,19 +262,90 @@ pub struct FunctionTypeId {
+        let FunctionType { symbol_id, .. } = *db.type_store()?.get_function(self);
+        let scope_id = symbol_id.symbol(&table).scope_id();
+        let scope = scope_id.scope(&table);
+        match scope.kind() {
+            ScopeKind::Class => {
+                if let Some(def) = scope.definition() {
+                    if let Some(symbol_id) = scope.defining_symbol() {
+                        if let Type::Class(class) =
+                            db.infer_definition_type(self.file_id, symbol_id, def)?
+                        {
+                            Ok(Some(class))
+                        } else {
+                            Ok(None)
+                        }
+                    } else {
+                        Ok(None)
+                    }
+                } else {
+                    Ok(None)
+                }
+            }
+            _ => Ok(None),
+        }
```
Code-style nit: I'd be inclined to "unspool" this a bit:

```suggestion
        if !matches!(scope.kind(), ScopeKind::Class) {
            return Ok(None);
        }
        let Some(def) = scope.definition() else {
            return Ok(None);
        };
        let Some(symbol_id) = scope.defining_symbol() else {
            return Ok(None);
        };
	let Type::Class(class) = db.infer_definition_type(self.file_id, symbol_id, def)? else {
	    return Ok(None);
	};
```

AlexWaygood commented on crates/red_knot/src/lint.rs:
```diff
@@ -140,6 +142,66 @@ fn lint_unresolved_imports(context: &SemanticLintContext) -> QueryResult<()> {
+            .db
+            .infer_definition_type(context.file_id, symbol, definition.clone())?;
+        let Type::Function(func) = ty else {
+            debug_assert!(false, "type of a FunctionDef should always be a Function");
+            continue;
```
What's the motivation here for only panicking in a debug build? Ideally I feel like we wouldn't have to have an assertion like this, but if we do, I'd probably do something like this instead:

```suggestion
            unreachable!("type of a FunctionDef should always be a Function");
```

carljm replied:
Yeah, I wasn't sure about this, it's probably best to just use `unreachable`. Or even better to wrap this up in an API so most code doesn't have to make any assertion like this.

AlexWaygood commented on crates/red_knot/src/lint.rs:
```diff
@@ -140,6 +142,66 @@ fn lint_unresolved_imports(context: &SemanticLintContext) -> QueryResult<()> {
+            // not a method of a class
+            continue;
+        };
+        let mut override_decorated = false;
+        for deco_ty in func.function(context.db)?.decorators() {
+            let Type::Function(deco_func) = deco_ty else {
+                continue;
+            };
+            if deco_func.file() == typing_file && deco_func.symbol(context.db)? == typing_override {
+                override_decorated = true;
+                break;
+            }
+        }
```
This is slightly more verbose but feels slightly cleaner to me -- but it's just a stylistic preference, feel free to ignore

```suggestion
        let override_decorated = func
            .function(context.db)?
            .decorators()
            .iter()
            .filter_map(|deco_type| {
                if let Type::Function(deco_func) = deco_type {
                    Some(deco_func)
                } else {
                    None
                }
            })
            .any(|deco_func| {
                deco_func.file() == typing_file
                    && deco_func
                        .symbol(context.db)
                        .is_ok_and(|symbol| symbol == typing_override)
            });
```

carljm replied:
I played around with options like this for a while (including this very one), but I don't think it's good to just ignore an error result from the `.symbol(context.db)` call like this. The current code bubbles it up to the caller.

AlexWaygood replied:
That makes sense. You could `.collect()` them into a single `Result<Vec<_>>` like this, and then bubble up the error to the caller:

```suggestion
        let override_decorated = func
            .function(context.db)?
            .decorators()
            .iter()
            .filter_map(|deco_type| {
                let Type::Function(deco_func) = deco_type else {
                    return None;
                };
                if deco_func.file() != typing_file {
                    return None;
                }
                Some(deco_func.symbol(context.db))
            })
            .collect::<QueryResult<Vec<SymbolId>>>()?
            .iter()
            .any(|symbol_id| *symbol_id == typing_override);
```

AlexWaygood replied:
Maybe something like this would be best -- it's pretty similar to your current code, but the `override_decorated` variable doesn't have to be marked as mutable in the main `for` loop, which I think is the main thing I didn't really like about your current code

```suggestion
        let override_decorated = {
            let mut found_override_decorator = false;
            for deco_ty in func.function(context.db)?.decorators() {
                let Type::Function(deco_func) = deco_ty else {
                    continue;
                };
                if deco_func.file() == typing_file
                    && deco_func.symbol(context.db)? == typing_override
                {
                    found_override_decorator = true;
                    break;
                }
            }
            found_override_decorator
        };
```

carljm replied:
Yeah, I'd also considered the collect-to-a-Vec option, but allocating unnecessarily didn't seem worth it. The `try_fold` option is nice! And yeah I agree this should probably just be a `has_decorator` function.

MichaReiser reviewed (commented):
Nice that you managed to get this rule working for today!

I think this PR shows a few issues that we need to address:

* `SemanticLintContext` probably must be generic over `Db` to prevent that we need to leak the internal storage layout on the `Database` API. We should solve this as part of this PR.
* Our Database is somewhat awkward because it requires setting the right `HasJar` and `SemanticDb` type bounds. I think salsa doesn't have this problem because it can fallback to the `HasJarDyn` trait to resolve an ingredient. We need a better solution for that, but that's out of scope for this PR
* My ultimate goal was to prevent that lint rules get access to `db` and instead force them to go through the context. I can see now how this is challenging with e.g. `FunctionTypeId::name` taking a `db` as argument. I still think that not giving lint rules access to the `db` is desired, but I'm not aware of a good solution right now. 
* This is related to not giving access to `db` to the semantic lint rules. I think that it's very easy today to mix local ids from different files. Ideally, lint rules would only ever get local ids for the current file or otherwise global ids.

carljm commented on crates/red_knot/src/program/mod.rs:
```diff
@@ -110,10 +114,23 @@ impl SemanticDb for Program {
         symbol_table(self, file_id)
     }
 
+    fn type_store(&self) -> QueryResult<&crate::types::TypeStore> {
+        type_store(self)
+    }
+
```
Yeah I agree, in general with this PR I was just trying to get the rule working however was fastest, but really all the uses of `type_store` should become their own queries.

carljm commented on crates/red_knot/src/symbols.rs:
```diff
@@ -67,7 +67,10 @@ pub(crate) enum ScopeKind {
-    child_scopes: Vec<ScopeId>,
+    parent: Option<ScopeId>,
+    children: Vec<ScopeId>,
+    definition: Option<Definition>,
+    defining_symbol: Option<SymbolId>,
```
Yes I can add docs for these.

We cannot navigate from `SymbolId` to `Definition` because a symbol can have multiple definitions, and we need to know which specific definition. This is a general problem that affected a lot of things in the design of this PR.

carljm replied:
Yeah, that might make sense. I'll add a TODO and leave that for another PR though.

carljm commented on crates/red_knot/src/symbols.rs:
```diff
@@ -80,6 +83,14 @@ impl Scope {
     }
+
+    pub(crate) fn definition(&self) -> Option<Definition> {
+        self.definition.clone()
```
I think we discussed this in comments on a previous PR. I don't see any reason `Definition` should ever become large; all of the information we might ever want to pass through it is already available on the AST node, and `Definition` should just be a thin wrapper around a `TypedNodeKey`.

In fact I would like to make `Definition` `Copy`, but I ran into some difficulty with doing that -- I can't remember what the difficulty was anymore.

If searching for `TypedNodeKey` becomes a perf bottleneck, we will have to reconsider the design here. Short of a directly-referenceable (i.e. Arc) AST, I don't see many great options. We can copy all the information we need out of the AST, but that's a lot of extra allocation (and then `Definition` will become quite large indeed).

carljm replied:
I think the idea of interning Definitions is probably a good one. Adding a TODO for that.

carljm commented on crates/red_knot/src/symbols.rs:
```diff
@@ -410,6 +457,27 @@ where
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some(scope_id) = self.scope_id {
+            let parent = self.table.scopes_by_id[scope_id].parent;
```
Maybe. Or maybe we should have `scope_id.parent_scope(&table)`. I'm generally not sure which style API makes more sense.

MichaReiser replied:
Yeah, I don't have a good answer to this yet.

carljm commented on crates/red_knot/src/symbols.rs:
```diff
@@ -410,6 +457,27 @@ where
+
+impl<'a> FusedIterator for ParentScopeIdsIterator<'a> {}
+
 pub(crate) struct ScopeIterator<'a, I> {
```
It used to return `Scope` objects, now it returns `(ScopeId, Scope)` pairs, since a `Scope` doesn't know its own id. I think this is a pretty natural pair and I don't really want to introduce a new struct for it, but I could document it.

Maybe we don't need this iterator at all if we just implement all access of `Scope` data via methods exposed on `ScopeId` that take a `SymbolTable`. Then `Scope` becomes an internal implementation detail. The advantage here is that you always just work with `ScopeId` and don't worry about `ScopeId` vs `Scope`. The disadvantage is that you can only access data one field at a time, and if you need multiple you end up doing multiple lookups. But considering these lookups are just indexed access into an `IndexVec`, maybe they are so cheap this doesn't matter.

carljm commented on crates/red_knot/src/lint.rs:
```diff
@@ -140,6 +142,66 @@ fn lint_unresolved_imports(context: &SemanticLintContext) -> QueryResult<()> {
+        return Ok(());
+    };
+    let typing_file = context.db.module_to_file(typing_module)?;
+    let typing_table = context.db.symbol_table(typing_file)?;
```
Yes, I think that's a good direction to move the API.

I want to avoid over-fitting our API to our first few lint rules, so it was kind of intentional (also just that I was short on time) that I didn't do much API design thinking in this PR, just pushed to get it working. I want to abstract things when it is clear we have multiple needs for them. But I think this is a pretty clear case of something we will have many uses for.

MichaReiser replied:
yeah, that sounds reasonable to me. It might be an opportunity for Patrick to work out an API for this as part of the work he does on the new lint rule.

carljm replied:
I added a `resolve_global_symbol` query to encapsulate this.

carljm commented on crates/red_knot/src/lint.rs:
```diff
@@ -140,6 +142,66 @@ fn lint_unresolved_imports(context: &SemanticLintContext) -> QueryResult<()> {
+        }
+        let ty = context
+            .db
+            .infer_definition_type(context.file_id, symbol, definition.clone())?;
```
We can't retrieve definition from symbol because the relationship is one(symbol)-to-many(definitions).

I would like to solve this issue by making `Definition` `Copy`, if I can resolve whatever problem stopped me from doing that already.

I think we could also pass `Definition` by reference here. Whether `Copy` or reference makes more sense depends on whether we expect `Definition` to get any bigger, which I don't.

carljm replied:
I think interning Definitions is probably a good direction here; I added a TODO for that.

carljm commented on crates/red_knot/src/lint.rs:
```diff
@@ -140,6 +142,66 @@ fn lint_unresolved_imports(context: &SemanticLintContext) -> QueryResult<()> {
+            let Type::Function(deco_func) = deco_ty else {
+                continue;
+            };
+            if deco_func.file() == typing_file && deco_func.symbol(context.db)? == typing_override {
```
Yes, I agree this is a good direction to push the API.

I think we can provide APIs that prevent a lint rule from _needing_ to get another file's symbol table. I'm not sure how we prevent a lint rule from being _able_ to do that, though. It seems like it will require not giving lint rules direct db access at all. Which may be feasible but will require a lot of duplication of parts of db API surface on lint-context.

carljm replied:
I added a GlobalSymbolId type and a `resolve_global_symbol` API that returns it, as well as changing other queries (`infer_symbol_type`, `infer_definition_type`) to take it.

We still use local SymbolIds some places, but fewer now.

This doesn't really do anything to make it harder for a lint rule to get access to some other module's symbol table, but it provides an alternative API that should mean they don't need to.

carljm commented:
Thanks for the review!

> SemanticLintContext probably must be generic over Db to prevent that we need to leak the internal storage layout on the Database API. We should solve this as part of this PR.

Why does using a trait object db require us to leak the internal storage layout on the database API? I think it led to that in this PR just because I was cutting corners to get things working, but I don't think it requires it. Instead it requires that all data access we need available to lint rules must be implemented as queries on the Db trait interface. These queries don't need to leak implementation details, but it will mean there will be a lot of them.

In some ways it seems like making everything generic over Db has more risk of leaking internals, because it means that the full `Db` interface is always available instead of just what we expose on the trait.

(I don't have strong opinions here, I'm still working out the pros and cons of different patterns and trying to understand the reasoning for different options.)

> My ultimate goal was to prevent that lint rules get access to db and instead force them to go through the context. I can see now how this is challenging with e.g. FunctionTypeId::name taking a db as argument. I still think that not giving lint rules access to the db is desired, but I'm not aware of a good solution right now.

Yes, I think this will be pretty challenging with the interning-based storage approach, because pretty much anything you might want to do requires a db.

> This is related to not giving access to db to the semantic lint rules. I think that it's very easy today to mix local ids from different files. Ideally, lint rules would only ever get local ids for the current file or otherwise global ids.

Yeah, I agree we should try to avoid this footgun, though it's not clear to me how to actually make it impossible. (Even a global ID will contain local IDs, and making those inaccessible will make those global IDs quite difficult or impossible to use in the code that actually needs to break them apart to do the queries. Maybe we can address this with crate visibility boundaries, if in future we break lint rules into a separate crate from the db.)

carljm commented:
Addressed most review comments. Would like to discuss this more, as the right path isn't clear to me:

> SemanticLintContext probably must be generic over Db to prevent that we need to leak the internal storage layout on the Database API. We should solve this as part of this PR.

MichaReiser commented on crates/red_knot/src/symbols.rs:
```diff
@@ -33,6 +32,32 @@ pub fn symbol_table(db: &dyn SemanticDb, file_id: FileId) -> QueryResult<Arc<Sym
+    }))
+}
+
+#[derive(Copy, Clone, Debug, PartialEq)]
```
I think it's okay to add  `Eq` as well.

```suggestion
#[derive(Copy, Clone, Debug, PartialEq, Eq)]
```
